marvin/core/test_runner.py

```diff
@@ -0,0 +1,79 @@
+import sys
+
+from marvin.data.data_providers.null_data_provider import NullDataProvider
+from marvin.core.status import Status
+from marvin.exceptions import ContextSkippedException
+from marvin.report.events import TestStartedEvent, TestEndedEvent, TestSetupStartedEvent, TestSetupEndedEvent, \
+    TestIterationStartedEvent, TestIterationEndedEvent, TestTearDownStartedEvent, TestTearDownEndedEvent
+from marvin.util import NO_EXCEPTION
+
+
+class TestRunner(object):
+    """
+    Internal Marvin Test executor. Implement the logic of a test execution flow, signals, events, etc
+    """
+
+    def __init__(self, test_script, data_provider):
+        self._test = test_script
+        self._data = data_provider or NullDataProvider()
+        self._status = None
+        self._exceptions = []
+        self._skip_iteration = False
+        self._skip_teardown = False
+
+    def execute(self):
+        test_started = TestStartedEvent(self._test, self._data)
+        self._test.publisher.notify(test_started)
+
+        self._execute()
+
+        test_ended = TestEndedEvent(self._test, test_started.timestamp, self._status, self._exceptions)
+        self._test.publisher.notify(test_ended)
+        self._test.ctx.sub_context_finished(self._status)
+
+    def _execute(self):
+        self._do_block('setup', self._data.setup_data(), TestSetupStartedEvent, TestSetupEndedEvent)
+
+        for it_data in self._data.iteration_data():
+            self._do_block('run', it_data, TestIterationStartedEvent, TestIterationEndedEvent)
+
+        self._do_block('tear_down', self._data.tear_down_data(), TestTearDownStartedEvent, TestTearDownEndedEvent)
+
+    def _do_block(self, block_type, data, started_event_class, ended_event_class):
+        if self._should_skip_block(block_type):
```

Comment:
The setup phase, each of the iterations, are the tear down phase are blocks. This line is to decide whether the block should be executed or not. The current rules (which we can discuss and change if necessary) are:
  * The setup block is always executed (i.e. this call will always return False for `block_type='setup'`).
  * Iteration blocks are not executed if the setup phase resulted in FAIL, or SKIP.
  * The destroy block is not executed only if the setup phase resulted in SKIP (I'm not completely if this makes sense though)